Mathlib/Logic/Hydra.lean

```diff
@@ -42,122 +42,118 @@ open Multiset Prod
 
 variable {α : Type _}
 
-/-- The relation that specifies valid moves in our hydra game. `cut_expand r s' s`
+/-- The relation that specifies valid moves in our hydra game. `CutExpand r s' s`
   means that `s'` is obtained by removing one head `a ∈ s` and adding back an arbitrary
   multiset `t` of heads such that all `a' ∈ t` satisfy `r a' a`.
 
-  This is most directly translated into `s' = s.erase a + t`, but `multiset.erase` requires
-  `decidable_eq α`, so we use the equivalent condition `s' + {a} = s + t` instead, which
+  This is most directly translated into `s' = s.erase a + t`, but `Multiset.erase` requires
+  `DecidableEq α`, so we use the equivalent condition `s' + {a} = s + t` instead, which
   is also easier to verify for explicit multisets `s'`, `s` and `t`.
 
   We also don't include the condition `a ∈ s` because `s' + {a} = s + t` already
   guarantees `a ∈ s + t`, and if `r` is irreflexive then `a ∉ t`, which is the
   case when `r` is well-founded, the case we are primarily interested in.
 
-  The lemma `relation.cut_expand_iff` below converts between this convenient definition
+  The lemma `Relation.cutExpand_iff` below converts between this convenient definition
   and the direct translation when `r` is irreflexive. -/
 def CutExpand (r : α → α → Prop) (s' s : Multiset α) : Prop :=
-  ∃ (t : Multiset α)(a : α), (∀ a' ∈ t, r a' a) ∧ s' + {a} = s + t
+  ∃ (t : Multiset α) (a : α), (∀ a' ∈ t, r a' a) ∧ s' + {a} = s + t
 #align relation.cut_expand Relation.CutExpand
 
 variable {r : α → α → Prop}
 
-theorem cutExpand_le_invImage_lex [hi : IsIrrefl α r] :
-    CutExpand r ≤ InvImage (Finsupp.Lex (rᶜ ⊓ (· ≠ ·)) (· < ·)) toFinsupp :=
-  fun s t ⟨u, a, hr, he⟩ => by
+theorem cutExpand_le_invImage_lex [IsIrrefl α r] :
+    CutExpand r ≤ InvImage (Finsupp.Lex (rᶜ ⊓ (· ≠ ·)) (· < ·)) toFinsupp := by
+  rintro s t ⟨u, a, hr, he⟩
+  replace hr := fun a' ↦ mt (hr a')
   classical
-    refine' ⟨a, fun b h => _, _⟩ <;> simp_rw [to_finsupp_apply]
-    · apply_fun count b  at he
-      simp_rw [count_add] at he
-      convert he <;> convert (add_zero _).symm <;> rw [count_eq_zero] <;> intro hb
-      exacts[h.2 (mem_singleton.1 hb), h.1 (hr b hb)]
-    · apply_fun count a  at he
-      simp_rw [count_add, count_singleton_self] at he
-      apply Nat.lt_of_succ_le
-      convert he.le
-      convert (add_zero _).symm
-      exact count_eq_zero.2 fun ha => hi.irrefl a <| hr a ha
+  refine ⟨a, fun b h ↦ ?_, ?_⟩ <;> simp_rw [toFinsupp_apply]
+  · replace he := congr_arg (count b) he -- porting note: was apply_fun count b  at he
```

Comment:
In fact, `apply_fun` has been improved in the meantime, and works here. (Also a few lines below.)